source/common/grpc/common.cc

```diff
@@ -23,23 +27,36 @@ void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string&
 
 Buffer::InstancePtr Common::serializeBody(const google::protobuf::Message& message) {
   // http://www.grpc.io/docs/guides/wire.html
+  // Reserve enough space for the entire message and the 5 byte header.
   Buffer::InstancePtr body(new Buffer::OwnedImpl());
-  uint8_t compressed = 0;
-  body->add(&compressed, sizeof(compressed));
-  uint32_t size = htonl(message.ByteSize());
-  body->add(&size, sizeof(size));
-  body->add(message.SerializeAsString());
-
+  const uint32_t size = message.ByteSize();
+  const uint32_t alloc_size = size + 5;
+  Buffer::RawSlice iovec;
+  body->reserve(alloc_size, &iovec, 1);
+  ASSERT(iovec.len_ >= alloc_size);
+  iovec.len_ = alloc_size;
+  uint8_t* current = reinterpret_cast<uint8_t*>(iovec.mem_);
+
+  *current++ = 0; // flags
```

Comment:
use https://github.com/lyft/envoy/blob/master/source/common/grpc/codec.h#L27 to do this? cc @fengli79